subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2011 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.plugins.ide.eclipse.model;
+
+import com.google.common.base.Objects;
+import groovy.lang.Closure;
+import groovy.lang.DelegatesTo;
+import org.gradle.api.Action;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.internal.ClosureBackedAction;
+
+/**
+ * The gradle DSL model of an Eclipse resource filter.
+ *
+ * Later this will be extended with additional helpers, but for now it just allows specifying a filter with a custom matcher
+ * and configuring whether it is an include/exclude filter that applies to files, folders, or both.  This example excludes the 'node_modules' folder.
+ *
+ * <pre autoTested=''>
+ * apply plugin: 'java'
+ * apply plugin: 'eclipse'
+ *
+ * eclipse {
+ *   project {
+ *     // the following closure configures the ResourceFilter
+ *     resourceFilter {
+ *       appliesTo = 'FOLDERS'
+ *       type = 'EXCLUDE_ALL'
+ *       matcher {
+ *         id = 'org.eclipse.ui.ide.multiFilter'
+ *         arguments = '1.0-name-matches-false-false-node_modules'
+ *       }
+ *     }
+ *   }
+ * }
+ * </pre>
+ *
+ * @since NEXT_VERSION
+ */
+public final class ResourceFilter {
+    private ResourceFilterAppliesTo appliesTo = ResourceFilterAppliesTo.FILES_AND_FOLDERS;
+    private ResourceFilterType type = ResourceFilterType.EXCLUDE_ALL;
+    private boolean recursive = true;
+    private ResourceFilterMatcher matcher;
+
+    // Needed for DSL configuration Actions
```

Comment:
Yes, we can't validate this up-front, we can only give reasonable defaults (or fail later on if the user forgot something)